contracts/marketpools.js

```diff
@@ -0,0 +1,350 @@
+    await api.db.createTable('params');
+
+    const params = {};
+    params.poolCreationFee = '1000';
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  const { poolCreationFee } = payload;
+  if (api.sender !== api.owner) return;
+  const params = await api.db.findOne('params', {});
+  if (poolCreationFee) {
+    if (!api.assert(typeof poolCreationFee === 'string' && !api.BigNumber(poolCreationFee).isNaN() && api.BigNumber(poolCreationFee).gte(0), 'invalid poolCreationFee')) return;
+    params.poolCreationFee = poolCreationFee;
+  }
+  await api.db.update('params', params);
+};
+
+function getAmountIn(amountOut, liquidityIn, liquidityOut) {
+  if (!api.assert(api.BigNumber(amountOut).gt(0), 'insufficient output amount')
+    || !api.assert(api.BigNumber(liquidityIn).gt(0)
+      && api.BigNumber(liquidityOut).gt(0)
+      && api.BigNumber(amountOut).lt(liquidityOut), 'insufficient liquidity')) return false;
+  const num = api.BigNumber(liquidityIn).times(amountOut);
+  const den = api.BigNumber(liquidityOut).minus(amountOut);
+  return num.dividedBy(den);
+}
+
+function getAmountOut(amountIn, liquidityIn, liquidityOut) {
+  if (!api.assert(api.BigNumber(amountIn).gt(0), 'insufficient output amount')
+    || !api.assert(api.BigNumber(liquidityIn).gt(0)
+      && api.BigNumber(liquidityOut).gt(0), 'insufficient liquidity')) return false;
+  const num = api.BigNumber(amountIn).times(liquidityOut);
+  const den = api.BigNumber(liquidityIn).plus(amountIn);
+  const amountOut = num.dividedBy(den);
+  if (!api.assert(api.BigNumber(amountOut).lt(liquidityOut), 'insufficient liquidity')) return false;
+  return amountOut;
+}
+
+async function validateOracle(pool, newPrice, maxDeviation = api.BigNumber('0.01')) {
+  const [baseSymbol, quoteSymbol] = pool.tokenPair.split(':');
+  // eslint-disable-next-line no-template-curly-in-string
+  const baseMetrics = baseSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: baseSymbol })
+    : { lastPrice: 1 };
+  // eslint-disable-next-line no-template-curly-in-string
+  const quoteMetrics = quoteSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: quoteSymbol })
+    : { lastPrice: 1 };
+  if (!baseMetrics || !quoteMetrics) return null; // no oracle available
+  const oracle = api.BigNumber(baseMetrics.lastPrice).dividedBy(quoteMetrics.lastPrice);
+  const dev = api.BigNumber(newPrice).minus(oracle).abs().dividedBy(oracle);
+  // api.debug(`${oracle} -> ${dev} / ${maxDeviation}`);
+  if (!api.assert(api.BigNumber(dev).lte(maxDeviation), 'exceeded max deviation from order book')) return false;
+  return true;
+}
+
+function validateLiquiditySwap(pool, baseDelta, quoteDelta) {
+  const p = api.BigNumber(pool.quoteQuantity).dividedBy(pool.baseQuantity).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
```

Comment:
for price comparison, ROUND_HALF_UP  for these two. (P and P')

Reply:
Agreed, changed.